### ISSUE TITLE
Fix Hardware Counters diagnoser

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/BenchmarkDotNet#1288]
I've tried to use Hardware Counters diagnoser today and I've failed to get it working.

The person who would like to fix it should most probably:

1. Make sure that this is not an OS bug. I am using preview Windows builds and it can be just an OS bug.

Try running a "Hello World" example:

Install both our packages from the [CI feed](https://benchmarkdotnet.org/articles/guides/nuget.html#nightly):

```xml
<packageSources>
  <add key="bdn-ci" value="https://ci.appveyor.com/nuget/benchmarkdotnet" />
</packageSources>
```

```xml
<ItemGroup>
  <PackageReference Include="BenchmarkDotNet" Version="0.11.5.1197" />
  <PackageReference Include="BenchmarkDotNet.Diagnostics.Windows" Version="0.11.5.1197" />
</ItemGroup>
```

Run the simplest app:

```cs
static void Main(string[] args) => BenchmarkRunner.Run<Repro>();

[HardwareCounters(HardwareCounter.InstructionRetired)]
public class Repro
{
    public int Simple()
    {
        int result = 0;
        for (int i = 0; i < 100_000; i++)
        {
            result ^= i;
        }
        return result;
    }
}
```

If you get any number in `InstructionRetired/Op` column, then it's a bug in OS version that I am using. If so, please just report the Windows version number printed by BenchmarkDotNet (example: `BenchmarkDotNet=v0.11.5.1159-nightly, OS=Windows 10.0.18362`)

2. Downgrade the referenced version of [TraceEvent](https://www.nuget.org/packages/Microsoft.Diagnostics.Tracing.TraceEvent/) library and see if it's a regression in TraceEvent (build BenchmarkDotNet from source to make sure you are using the old version) or a bug in BenchmarkDotNet

3. Create a small "Hello World" example using TraceEvent based on my [old blog post](https://adamsitnik.com/Hardware-Counters-ETW/), if it works fine then it's a BenchmarkDotNet bug

4. When you fix the bug in BenchmarkDotNet|TraceEvent, please test [Hardware Counters with Dissasembly Diagnoser](https://adamsitnik.com/Disassembly-Diagnoser/#the-ultimate-combination) - I guess that it does not work as well

**Comment by terrance00** [comment]
Hi, I've forked and am working on this.

**Comment by adamsitnik** [comment]
@terrance00 great! please let me know when something is not clear

**Comment by terrance00** [comment]
I assumed we actually needed a `Benchmark` attribute on the `Simple` method.

I added a smaller iteration count, because I have no patience.

And ran the tests for a .net core and .net framework console app.

Output from .net framework:
```
BenchmarkDotNet=v0.11.5.1197-nightly, OS=Windows 10.0.17134.1069 (1803/April2018Update/Redstone4)
Intel Core i7-4790 CPU 3.60GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4018.0), X86 LegacyJIT  [AttachedDebugger]
  Job-EQFHPC : .NET Framework 4.8 (4.8.4018.0), X86 LegacyJIT

IterationCount=5

| Method |      Mean |     Error |    StdDev | InstructionRetired/Op |
|------- |----------:|----------:|----------:|----------------------:|
| Simple | 0.0329 ns | 0.0440 ns | 0.0114 ns |                     3 |
```

Output from .net core:
```
// * Summary *

BenchmarkDotNet=v0.11.5.1197-nightly, OS=Windows 10.0.17134.1069 (1803/April2018Update/Redstone4)
Intel Core i7-4790 CPU 3.60GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100
  [Host]     : .NET Core 2.2.5 (CoreCLR 4.6.27617.05, CoreFX 4.6.27618.01), X64 RyuJIT  [AttachedDebugger]
  Job-WHAYKX : .NET Core 2.2.5 (CoreCLR 4.6.27617.05, CoreFX 4.6.27618.01), X64 RyuJIT

IterationCount=5

| Method |      Mean |     Error |    StdDev | InstructionRetired/Op |
|------- |----------:|----------:|----------:|----------------------:|
| Simple | 0.7084 ns | 0.0711 ns | 0.0185 ns |                     3 |
```

So, seems like it might be a bug on your side? I noticed the dotnet preview version you were using as well - perhaps try using a released version.

Anything I missed?

I'd like to contribute  - just keep in mind my noobness at this awesome project :)

**Comment by adamsitnik** [comment]
@terrance00 great, thanks for trying it. It means that something is wrong on my side (either Windows bug or hardware issue)

Could you please test [Hardware Counters with Dissasembly Diagnoser](https://adamsitnik.com/Disassembly-Diagnoser/#the-ultimate-combination)? I guess that it got broken some time ago.

```cs
[HardwareCounters(HardwareCounter.BranchMispredictions, HardwareCounter.BranchInstructions)]
[DisassemblyDiagnoser(printAsm: true, printSource: true)]
public class Cpu_BranchPerdictor
{
    private const int N = 32767;
    private readonly int[] sorted, unsorted;

    public Cpu_BranchPerdictor()
    {
        var random = new Random(0);
        unsorted = new int[N];
        sorted = new int[N];
        for (int i = 0; i < N; i++)
            sorted[i] = unsorted[i] = random.Next(256);
        Array.Sort(sorted);
    }

    private static int Branch(int[] data)
    {
        int sum = 0;
        for (int i = 0; i < N; i++)
            if (data[i] >= 128)
                sum += data[i];
        return sum;
    }

    [Benchmark]
    public int SortedBranch() => Branch(sorted);

    [Benchmark]
    public int UnsortedBranch() => Branch(unsorted);
}
```

Such config (using both HadrwareCounters diagnoser and DissasemblyDiagnoser) should enable BDN to use [InstructionPointerExporter](https://github.com/dotnet/BenchmarkDotNet/blob/a76f438a81e3e0235a2e9b1249a5280038189689/src/BenchmarkDotNet/Exporters/InstructionPointerExporter.cs) and export html file with hardware counter information per machine instruction

![obraz](https://user-images.githubusercontent.com/6011991/67426813-29d55300-f5db-11e9-8669-9e9cb722a30a.png)

**Comment by adamsitnik** [comment]
I've split this issue into few and sent the fixes in #1547 #1548 and #1550 Now everything works like a charm
